File: haskell/euler_15.hs
-- This calculates a level in Pacal's triangle based on the last
-- level.
next_level :: [Integer] -> [Integer]
next_level l = [1] ++ [l !! n  + l !! (n + 1) | n <- [0..(length l) - 2]] ++ [1]
                                                       
-- List of the levels of Pascal's triangle
pascals_triangle = [1] : [next_level (pascals_triangle !! n) | n <- [0..]]

-- calculate the number of paths through a lattice with sides of a
-- given size.
num_paths :: Int -> Integer
num_paths size = 
  let
    level = pascals_triangle !! (size * 2)
    middle_idx = div (length level) 2
  in
   level !! middle_idx

main = do
  print $ num_paths 20

File: python/euler_15.py
# paths through a lattice

def pascals_triangle(depth):
    rval = [0] * depth
    rval[0] = 1
    work = rval[:]
    for i in range(1, depth):
        temp = work
        work = rval
        rval = temp
        for j in range(1, depth):
            rval[j] = work[j - 1] + work[j]
    return rval

pvals = pascals_triangle(41)
print pvals[19]
print pvals[20]
print pvals[21]



